Python/instrumentation.c

```diff
@@ -0,0 +1,1986 @@
+    Py_DECREF(to_obj);
+    return err;
+}
+
+static void
+call_instrumentation_vector_protected(
+    PyThreadState *tstate, int event,
+    _PyInterpreterFrame *frame, _Py_CODEUNIT *instr, Py_ssize_t nargs, PyObject *args[])
+{
+    assert(_PyErr_Occurred(tstate));
+    PyObject *exc = _PyErr_GetRaisedException(tstate);
+    int err = call_instrumentation_vector(tstate, event, frame, instr, nargs, args);
+    if (err) {
+        Py_XDECREF(exc);
+    }
+    else {
+        _PyErr_SetRaisedException(tstate, exc);
+    }
+    assert(_PyErr_Occurred(tstate));
+}
+
+void
+_Py_call_instrumentation_exc0(
+    PyThreadState *tstate, int event,
+    _PyInterpreterFrame *frame, _Py_CODEUNIT *instr)
+{
+    assert(_PyErr_Occurred(tstate));
+    PyObject *args[3] = { NULL, NULL, NULL };
+    call_instrumentation_vector_protected(tstate, event, frame, instr, 2, args);
+}
+
+void
+_Py_call_instrumentation_exc2(
+    PyThreadState *tstate, int event,
+    _PyInterpreterFrame *frame, _Py_CODEUNIT *instr, PyObject *arg0, PyObject *arg1)
+{
+    assert(_PyErr_Occurred(tstate));
+    PyObject *args[5] = { NULL, NULL, NULL, arg0, arg1 };
+    call_instrumentation_vector_protected(tstate, event, frame, instr, 4, args);
+}
+
+
+int
+_Py_Instrumentation_GetLine(PyCodeObject *code, int index)
+{
+    _PyCoMonitoringData *monitoring = code->_co_monitoring;
+    assert(monitoring != NULL);
+    assert(monitoring->lines != NULL);
+    assert(index >= code->_co_firsttraceable);
+    assert(index < Py_SIZE(code));
+    _PyCoLineInstrumentationData *line_data = &monitoring->lines[index];
+    int8_t line_delta = line_data->line_delta;
+    int line = compute_line(code, index, line_delta);
+    return line;
+}
+
+int
+_Py_call_instrumentation_line(PyThreadState *tstate, _PyInterpreterFrame* frame, _Py_CODEUNIT *instr)
+{
+    frame->prev_instr = instr;
+    PyCodeObject *code = frame->f_code;
+    assert(is_version_up_to_date(code, tstate->interp));
+    assert(instrumentation_cross_checks(tstate->interp, code));
+    int i = instr - _PyCode_CODE(code);
+    _PyCoMonitoringData *monitoring = code->_co_monitoring;
+    _PyCoLineInstrumentationData *line_data = &monitoring->lines[i];
+    uint8_t original_opcode = line_data->original_opcode;
+    if (tstate->tracing) {
+        goto done;
+    }
+    PyInterpreterState *interp = tstate->interp;
+    int8_t line_delta = line_data->line_delta;
+    int line = compute_line(code, i, line_delta);
+    uint8_t tools = code->_co_monitoring->line_tools != NULL ?
+        code->_co_monitoring->line_tools[i] :
+        (interp->monitors.tools[PY_MONITORING_EVENT_LINE] |
+         code->_co_monitoring->local_monitors.tools[PY_MONITORING_EVENT_LINE]
+        );
+    PyObject *line_obj = PyLong_FromSsize_t(line);
+    if (line_obj == NULL) {
+        return -1;
+    }
+    PyObject *args[3] = { NULL, (PyObject *)code, line_obj };
+    while (tools) {
+        int tool = most_significant_bit(tools);
+        assert(tool < 8);
```

Comment:
Perhaps also assert `tool >= 0`, as in line 941?

```suggestion
        assert(tool >= 0 && tool < 8);
```